### PR TITLE
cli_parser: man-pages and help text generation using Homebrew::CLI::Parser

### PR DESCRIPTION
**Pull request by GauthamGoli**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----
https://github.com/Homebrew/brew/pull/4271#issuecomment-394177933
This PR is about generating `--help` text using `OptionParser` and to get feedback on the interface, and the way help text is being shown as its different from what we currently show.

`brew audit --help` would return something like this.

```
Usage: brew audit [options] [<formulae>]

Check <formulae> for Homebrew coding style violations. This should be
run before submitting a new formula.

If no <formulae> are provided, all of them are checked.
        --strict                     Run additional style checks, including Rubocop style checks.
        --online                     Run additional slower style checks that require a
                                     network connection.
        --new-formula                Run various additional style checks to determine if a new formula
                                     is eligible for Homebrew. This should be used when creating
                                     new formula and implies --strict and --online.
        --fix                        Fix style violations automatically using
                                     RuboCop's auto-correct feature.
        --display-cop-names          Include the RuboCop cop name for each violation
                                     in the output.
        --display-filename           Prefix everyline of output with name of the file or
                                     formula being audited, to make output easy to grep.
    -D, --audit-debug                Activates debugging and profiling
        --only                       Passing --only=method will run only the methods named audit_method,
                                     `method` should be a comma-separated list.
        --except                     Passing --except=method will run only the methods named audit_method,
                                     `method` should be a comma-separated list.
        --only-cops                  Passing --only-cops=cops will check for violations of only the listed
                                     RuboCop cops. `cops` should be a comma-separated list of cop names.
        --except-cops                Passing --except-cops=cops will skip checking the listed
                                     RuboCop cops violations. `cops` should be a comma-separated list of cop names.
    -v, --verbose                    Verbose mode.
    -d, --debug                      Display debug info.
    -h, --help                       Show this message
```

### CODE REVIEW

**Comment by MikeMcQuaid** [comment]
Can you run `brew man` so we can see the different help text output? Thanks!

File: Library/Homebrew/dev-cmd/audit.rb [hunk 1]
```diff
@@ -54,19 +54,27 @@ module Homebrew
...
-      switch      "--display-filename"
+      banner <<~EOS
+        Usage: brew audit [options] [<formulae>]
+
+        Check <formulae> for Homebrew coding style violations. This should be
+        run before submitting a new formula.
+
+        If no <formulae> are provided, all of them are checked.
+      EOS
+      switch      "--strict", description: "Run additional style checks, including Rubocop style checks."
```
[the remark this answers is not in the document]
Comment (GauthamGoli) [reply Library/Homebrew/dev-cmd/audit.rb]: `switch`'s first two arguments are already being specified for short and long options . eg: 
`switch   "-s", "--strict"`

[the remark this answers is not in the document]
Comment (GauthamGoli) [reply Library/Homebrew/dev-cmd/audit.rb]: May be we can separate them out by checking if they start with `-` and doesn't contain any  ` `

File: Library/Homebrew/brew.rb [hunk 2]
```diff
@@ -82,7 +82,7 @@
   # - a help flag is passed AND there is no command specified
   # - no arguments are passed
   # - if cmd is Cask, let Cask handle the help command instead
-  if (empty_argv || help_flag) && cmd != "cask"
+  if (empty_argv || help_flag) && cmd != "cask" && !internal_dev_cmd
```
[the remark this answers is not in the document]
Comment (MikeMcQuaid) [reply Library/Homebrew/brew.rb]: Also CC @reitermarkus as a note that it'd be good to get Cask's help behaving similar to the rest of Homebrew's.

**Comment by GauthamGoli** [comment]
I've been caught up at work.. will get back working on this in 3 days.

**Comment by GauthamGoli** [comment]
@MikeMcQuaid From what I see, `OptionParser` is very well suited for generating `brew <command> --help`

Man pages are formatted differently, we cannot directly use the help text as-is, but it CAN be used by writing some extra code. I looked at the man pages of `git` which also has sub-commands like `brew` 

Before proceeding, I have a question to clarify on help text rephrasing.
 

- `brew audit --help`  generated without `OptionParser`

```
brew audit [--strict] [--fix] [--online] [--new-formula] [--display-cop-names] [--display-filename] [--only=method|--except=method] [--only-cops=cops|--except-cops=cops] [formulae]:
    Check formulae for Homebrew coding style violations. This should be
    run before submitting a new formula.

    If no formulae are provided, all of them are checked.

    If --strict is passed, additional checks are run, including RuboCop
    style checks.
```
(snipped)
 - `brew audit --help` generated with `OptionParser`

```
brew audit [options] [<formulae>]

        Check <formulae> for Homebrew coding style violations. This should be
        run before submitting a new formula.

        If no <formulae> are provided, all of them are checked.

        --strict                     Run additional style checks, including Rubocop style checks.
```

Since `OptionParser` displays the description adjacent to the option, I rephrased it for brevity from 
```
If --strict is passed, additional checks are run, including RuboCop style checks.
``` 
to 
```
Run additional style checks, including Rubocop style checks.
```

Do we want to keep the current help text as-is , or rephrase it to latter ?

**Comment by MikeMcQuaid** [comment]
This doesn't look bad but would be nice to have some of the nicer formatting tweaks:

screenshot 2018-08-27 at 10 42 59

On a narrower terminal the existing version wraps a bit nicer:

screenshot 2018-08-27 at 10 43 52

**Comment by GauthamGoli** [comment]
I was facing formatting issues and had given up temporarily. I will give it a try again now to get the formatting right.

**Comment by MikeMcQuaid** [comment]
@GauthamGoli Thanks! We could always write our own generator if that'd be better.

**Comment by GauthamGoli** [comment]
@MikeMcQuaid Can you please generate manpage and let me know what you think?

File: Library/Homebrew/utils/formatter.rb [hunk 3]
```diff
@@ -31,6 +31,10 @@ def error(string, label: nil)
     label(label, string, :red)
   end
 
+  def wrap(s, width = 172, indentation = 0)
+    " "*indentation + s.gsub(/(.{1,#{width}})(\s+|\Z)/, "\\1\n").gsub("\n", "\n"+" "*indentation) + "\n"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/formatter.rb]: This is hard to follow. Could you use intermediate variables and/or string interpolation and put each `.gsub` on a new line?

File: Library/Homebrew/cli_parser.rb [hunk 4]
```diff
@@ -16,14 +16,28 @@ def initialize(&block)
...
+      end
+
+      def post_initialize
+        @parser.on_tail("-h", "--help", "Show this message") do
+          puts @parser
+        end
+      end
+
+      def wrap_option_desc(desc)
+        Formatter.wrap(desc, @desc_line_length).split("\n")
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cli_parser.rb]: What happens without these being wrapped? Is it the manpage/markdown/something else that looks weird?

File: Library/Homebrew/cli_parser.rb [hunk 5]
```diff
@@ -34,9 +48,18 @@ def switch(*names, description: nil, env: nil, required_for: nil, depends_on: ni
         enable_switch(*names) if !env.nil? && !ENV["HOMEBREW_#{env.to_s.upcase}"].nil?
       end
 
+      def banner(text)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cli_parser.rb]: Is the banner the help text? If so, a name relating to that might be more obvious.

File: Library/Homebrew/cli_parser.rb [hunk 6]
```diff
@@ -95,10 +128,10 @@ def enable_switch(*names)
       # These are common/global switches accessible throughout Homebrew
       def common_switch(name)
         case name
-        when :quiet   then [["-q", "--quiet"], :quiet]
-        when :verbose then [["-v", "--verbose"], :verbose]
-        when :debug   then [["-d", "--debug"], :debug]
-        when :force   then [["-f", "--force"], :force]
+        when :quiet   then [["-q", "--quiet"], :quiet, "Suppress warnings."]
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cli_parser.rb]: `Suppress any warnings`

File: Library/Homebrew/cli_parser.rb [hunk 7]
```diff
@@ -95,10 +128,10 @@ def enable_switch(*names)
...
+        when :verbose then [["-v", "--verbose"], :verbose, "Verbose mode."]
+        when :debug   then [["-d", "--debug"], :debug, "Display debug info."]
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cli_parser.rb]: `Display any debugging information`

File: Library/Homebrew/cli_parser.rb [hunk 8]
```diff
@@ -95,10 +128,10 @@ def enable_switch(*names)
       # These are common/global switches accessible throughout Homebrew
       def common_switch(name)
         case name
-        when :quiet   then [["-q", "--quiet"], :quiet]
-        when :verbose then [["-v", "--verbose"], :verbose]
-        when :debug   then [["-d", "--debug"], :debug]
-        when :force   then [["-f", "--force"], :force]
+        when :quiet   then [["-q", "--quiet"], :quiet, "Suppress warnings."]
+        when :verbose then [["-v", "--verbose"], :verbose, "Verbose mode."]
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cli_parser.rb]: `Make some output more verbose`

File: Library/Homebrew/cli_parser.rb [hunk 9]
```diff
@@ -95,10 +128,10 @@ def enable_switch(*names)
...
+        when :debug   then [["-d", "--debug"], :debug, "Display debug info."]
+        when :force   then [["-f", "--force"], :force, "Override any warnings/validations."]
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cli_parser.rb]: `Override warnings and enable potentially unsafe operations`

File: Library/Homebrew/dev-cmd/audit.rb [hunk 10]
```diff
@@ -52,22 +52,35 @@
...
+
+        Check <formulae> for Homebrew coding style violations. This should be
+        run before submitting a new formula.
+
+        If no <formulae> are provided, all of them are checked.
+
+      EOS
+      switch      "--strict", description: "Run additional style checks, including Rubocop style checks."
+      switch      "--online", description: "Run additional slower style checks that require a network connection."
+      switch      "--new-formula", description: "Run various additional style checks to determine if a new formula is eligible for Homebrew. This should be used when creating new formula and implies --strict and --online."
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/audit.rb]: This line (and some others) is a bit too long. Maybe stick each `description:` on a new line?

**Comment by MikeMcQuaid** [comment]
@GauthamGoli Good work on this! I think ideally if we're writing our own output I'd just aim for the output to be as close as possible to the existing rendered manpage output i.e. do stuff like `If X is passed,` and indent/wrap similarly to the existing. Anywhere where that seems to be fair too much work: shout here and we can talk about whether it's worth it.

File: Library/Homebrew/cli_parser.rb (hunk 5 above)
Comment (GauthamGoli) [reply Library/Homebrew/cli_parser.rb]: Banner is a brief description of the command shown above the options. 
Ex: 
```ruby
      banner <<~EOS
         Usage: brew audit [options] [<formulae>]
         Check <formulae> for Homebrew coding style violations. This should be
         run before submitting a new formula.
         If no <formulae> are provided, all of them are checked.
      EOS
```

Comment (MikeMcQuaid) [reply Library/Homebrew/cli_parser.rb]: @GauthamGoli Ok, I see. Maybe `usage_banner` or `help_banner` or similar?

File: Library/Homebrew/cli_parser.rb (hunk 4 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/cli_parser.rb]: @GauthamGoli I see, thanks 👍

**Comment by GauthamGoli** [comment]
@MikeMcQuaid The current manpage output for `brew audit` command is similar to the manpage of `git` and `nmap` (I like this format). I've noticed that a lot of other command line tools use this similar formatting in manpages instead of `If X is passed,` format. 

Do you think its worth transitioning to this new format ? If not, I'll make it as close to existing format.

**Comment by GauthamGoli** [comment]
@MikeMcQuaid  addressed the comments. Please take a look when you get time! 

Do we plan to remove the comments in the beginning of every `command.rb` that we previously used to generate help and manpages with?

File: Library/Homebrew/dev-cmd/audit.rb [hunk 11]
```diff
@@ -52,22 +52,55 @@
...
+        `audit` [<options>] [<formulae>]:
+
+        Check <formulae> for Homebrew coding style violations. This should be
+        run before submitting a new formula.
+
+        If no <formulae> are provided, all of them are checked.
+
+      EOS
+      switch      "--strict",            description: "Run additional style checks, "\
+                                                      "including Rubocop style checks."
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/audit.rb]: What about:

```ruby
switch      "--strict",
  description: "Run additional style checks, including Rubocop style checks."
```

**Comment by MikeMcQuaid** [comment]
> Do we plan to remove the comments in the beginning of every `command.rb` that we previously used to generate help and manpages with?

Provided e.g. `brew audit --help` works without it: yes.

> If this code style is OK, I'll do the same for other developer commands.

The manpage style is perfect. I agree it's actually better than the current style now 😍. There will be some slight inconsistency when the previous command style is used just for non-developer commands but I think it's worth it to get this merged.

File: Library/Homebrew/dev-cmd/audit.rb (hunk 11 above)
Comment (GauthamGoli) [reply Library/Homebrew/dev-cmd/audit.rb]: In the current style, one could see all the declared options in a single column (I think its more readable this way), but with above style the description would come in between. Its just a choice and I would defer the final call on this to you @MikeMcQuaid

Comment (MikeMcQuaid) [reply Library/Homebrew/dev-cmd/audit.rb]: I think it'd be preferable to be able to have the description on a single line, thanks 👍

File: Library/Homebrew/dev-cmd/audit.rb [hunk 12]
```diff
@@ -52,22 +52,53 @@
...
+  def audit_args
+    Homebrew::CLI::Parser.new do
+      usage_banner <<~EOS
+        `audit` [<options>] [<formulae>]:
+
+        Check <formulae> for Homebrew coding style violations. This should be
+        run before submitting a new formula.
+
+        If no <formulae> are provided, all of them are checked.
+
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/audit.rb]: Might be nice to make it so you don't need this trailing empty line?

File: Library/Homebrew/dev-cmd/audit.rb [hunk 13]
```diff
@@ -52,22 +52,53 @@
...
+      EOS
+      switch      "--strict",
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/audit.rb]: Sorry, terrible pedantry now but given the way `description` is I'd be tempted to say that `switch "--strict",` would look better.

**Comment by GauthamGoli** [comment]
@MikeMcQuaid   Now manpages for all dev commands are being generated through this. Can you once take a look, and let me know your thought? Thanks. 

There are minor hiccups like `--verbose` and `--debug` appearing for every sub-command.

**Comment by MikeMcQuaid** [comment]
screenshot 2018-09-30 at 10 56 11

@GauthamGoli I think it would be good if there was a bit more visual separation between commands either through indentation, keeping options on the same line as their first line of description or something else.


> There are minor hiccups like `--verbose` and `--debug` appearing for every sub-command.

Agreed, would be good to resolve this before merge. There could be a "global options" or similar section that apply to all commands.

**Comment by GauthamGoli** [comment]
@MikeMcQuaid  Is the indentation fine now? I made the command synopsis to be H3 ( ###) 
(but I'm not able to get the usual underline formatting to work in the sub heading. )

**Comment by MikeMcQuaid** [comment]
> Is the indentation fine now?

Looks great 😍.

More pedantry: 
screenshot 2018-10-01 at 14 26 16

- Ideally `[options]` would have only `options` in red (or maybe green? see `install-options` and not the `[ ]`
- Ideally `formulae` would be in `green` (and have the same `[]` fixes)
- As mentioned above: ideally `--verbose`, `--debug` wouldn't be repeated

After that: I think we're good to go. So close now! Great work again @GauthamGoli!

File: Library/Homebrew/dev-cmd/prof.rb [hunk 14]
```diff
@@ -1,7 +1,7 @@
-#:  * `prof` [<ruby options>]:
+#:  * `prof` [ruby options]:
 #:    Run Homebrew with the Ruby profiler.
 #:    For example:
-#       brew prof readall
+#:       brew prof readall
```
Comment (reitermarkus) [inline Library/Homebrew/dev-cmd/prof.rb]: Should be one fewer space here.

Comment (GauthamGoli) [reply Library/Homebrew/dev-cmd/prof.rb]: Nice catch :D thanks

File: Library/Homebrew/dev-cmd/audit.rb [hunk 15]
```diff
@@ -52,22 +52,51 @@
...
+      comma_array "--except",
+        description: "Passing `--except`=<method> will run only the methods named audit_<method>, "\
+                     "`method` should be a comma-separated list."
+      comma_array "--only-cops",
+        description: "Passing `--only-cops`=<cops> will check for violations of only the listed "\
+                     "RuboCop cops. `cops` should be a comma-separated list of cop names."
+      comma_array "--except-cops",
+        description: "Passing `--except-cops`=<cops> will skip checking the listed RuboCop cops "\
+                     "violations. `cops` should be a comma-separated list of cop names."
+      switch :verbose
```
Comment (reitermarkus) [inline Library/Homebrew/dev-cmd/audit.rb]: Is there a difference between `switch "--strict"` and `switch :verbose`? If not, we should only use either strings or symbols.

Comment (GauthamGoli) [reply Library/Homebrew/dev-cmd/audit.rb]: Symbols indicate that its a global option, and `CLI::Parser` would pick up predefined attributes related to that option.

**Comment by MikeMcQuaid** [comment]
> Now the command synopsis text i.e., `audit [options] [formulae]` also has formatting according to the review comments

Almost but not quite; `[options]` should be used when options are optional and `formulae` is correctly without `[]`. Otherwise: looking great 👍

File: Library/Homebrew/dev-cmd/man.rb [hunk 16]
```diff
@@ -118,6 +147,10 @@ def convert_man_page(markup, target)
       ronn.close_write
       ronn_output = ronn.read
       ronn_output.gsub!(%r{</var>`(?=[.!?,;:]?\s)}, "").gsub!(%r{</?var>}, "`") if format_flag == "--markdown"
+      unless format_flag == "--markdown"
+        ronn_output = ronn_output.gsub(/<code>/, "\\fB").gsub(%r{</code>}, "\\fR")
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/man.rb]: Put each `.gsub` on a new line and ideally document what each one is doing.

File: Library/Homebrew/dev-cmd/man.rb [hunk 17]
```diff
@@ -136,4 +169,56 @@ def target_path_to_format(target)
...
+
+  def generate_cmd_manpages(glob)
+    cmd_paths = Pathname.glob(glob)
+    man_page_lines = []
+    cmd_paths.each do |cmd_path|
+      begin
+        cmd_parser = Homebrew.send(cmd_arg_parser(cmd_path))
+        man_page_lines << cmd_manpage_lines(cmd_parser).join
+      rescue NoMethodError
+        man_page_lines << path_glob_commands(cmd_path.to_s)[0]
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/man.rb]: `.first`

File: Library/Homebrew/dev-cmd/man.rb (hunk 16 above)
Comment (GauthamGoli) [reply Library/Homebrew/dev-cmd/man.rb]: I've changed it so its self documented.

**Comment by GauthamGoli** [comment]
@MikeMcQuaid What about now? If all is fine now, I'll resolve the conflicts

**Comment by GauthamGoli** [comment]
@MikeMcQuaid any idea why the build is failing?

**Comment by MikeMcQuaid** [comment]
@GauthamGoli Fixing in https://github.com/Homebrew/homebrew-test-bot/pull/201 then will rerun.

**Comment by MikeMcQuaid** [comment]
Great work again @GauthamGoli 🎉
